x/incentives/keeper/distribution.go

```diff
@@ -0,0 +1,112 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/tharsis/evmos/x/incentives/types"
+)
+
+// Distribute transfers the allocated rewards to the participants of a given
+// incentive.
+//  - allocates the amount to be distribbuted from the inflaction pool
+//  - distributes the rewards to all paricpants
+//  - deletes all gas meters
+//  - sets the cumulative totalGas to zero
+//  - updates the remaining epochs of each incentive
+func (k Keeper) DistributeIncentives(ctx sdk.Context) error {
+	logger := k.Logger(ctx)
+
+	// Allocate rewards for each Incentive
+	coinsAllocated := k.allocateCoins(ctx)
+
+	k.IterateIncentives(
+		ctx,
+		func(incentive types.Incentive) (stop bool) {
+			// Distribute rewards and reset total gasMeter
+			k.rewardParticipants(ctx, incentive, coinsAllocated)
+			k.ResetTotalGas(ctx, incentive)
+
+			// Update Epoche and remove incentive from epoch if already finalized
+			incentive.Epochs--
+			if !incentive.IsActive() {
+				k.DeleteIncentive(ctx, incentive)
+				logger.Info(
+					"incentive finalized",
+					"contract", incentive.Contract,
+				)
+			}
+
+			return false
+		})
+
+	return nil
+}
+
+// Allocate amount of coins to be ditributed for each incentive
+func (k Keeper) allocateCoins(ctx sdk.Context) map[common.Address]sdk.Coins {
+	var coinsAllocated map[common.Address]sdk.Coins
+	moduleAddr := k.accountKeeper.GetModuleAddress(types.ModuleName)
+	k.IterateIncentives(
+		ctx,
+		func(incentive types.Incentive) (stop bool) {
+			var coins sdk.Coins
+			for _, allocation := range incentive.Allocations {
+				balance := k.bankKeeper.GetBalance(ctx, moduleAddr, allocation.Denom)
+				if !balance.IsPositive() {
+					continue
+				}
+				coinAllocated := balance.Amount.ToDec().Mul(allocation.Amount)
+				amount := coinAllocated.TruncateInt()
+				coin := sdk.NewCoin(allocation.Denom, amount)
```

Comment:
this way we don't run validation denom and amount validation each time
```suggestion
				coin := sdk.Coin{Denom: allocation.Denom, Amount: amount}
```